packages/core/src/sync-realtime/service.ts

```diff
@@ -8,878 +8,633 @@ import {
+      logFilters: realtimeSyncService.logFilterSources.map((l) => l.criteria),
+      factories: realtimeSyncService.factorySources.map((f) => f.criteria),
+      interval: {
+        startBlock: BigInt(realtimeSyncService.finalizedBlock.number + 1),
+        endBlock: BigInt(pendingFinalizedBlock.number),
+      },
     });
 
-    return false;
-  };
+    realtimeSyncService.finalizedBlock = pendingFinalizedBlock;
 
-  /**
-   * Check to see if re-requesting logs that we have locally will return a different result.
-   * If this occurs, treat it as a reorg.
-   */
-  private validateLogs = async (newFinalizedBlock: LightBlock) => {
-    this.common.logger.debug({
-      service: "realtime",
-      msg: `Validating local chain from block ${
-        this.finalizedBlock.number + 1
-      } to ${newFinalizedBlock.number} (network=${this.network.name})`,
-    });
+    // Note: This is where a finalization event would happen.
 
-    const _logs = await this._eth_getLogs({
-      fromBlock: numberToHex(this.finalizedBlock.number + 1),
-      toBlock: numberToHex(newFinalizedBlock.number),
+    realtimeSyncService.common.logger.debug({
+      service: "realtime",
+      msg: `Updated finality to block ${pendingFinalizedBlock.number} (network=${realtimeSyncService.network.name})`,
     });
-    const logs = sortLogs(_logs);
-
-    const matchedLogs = await this.getMatchedLogs(
-      logs,
-      BigInt(newFinalizedBlock.number),
-      false,
-    );
+  }
 
-    const localLogs = this.logs.filter(
-      (log) => log.blockNumber <= newFinalizedBlock.number,
-    );
+  realtimeSyncService.common.logger.debug({
+    service: "realtime",
+    msg: `Finished syncing new head block ${pendingLatestBlockNumber} (network=${realtimeSyncService.network.name})`,
+  });
 
-    // Evict local logs and blocks, emit reorg.
-    const handleLogsInconsistency = async (
-      localSafeBlockNumber: number | undefined,
-    ) => {
-      let commonAncestor: LightBlock;
-
-      if (localSafeBlockNumber === undefined) {
-        commonAncestor = this.finalizedBlock;
-      } else {
-        commonAncestor =
-          this.blocks.findLast((b) => b.number <= localSafeBlockNumber) ??
-          this.finalizedBlock;
-      }
+  return false;
+};
 
-      this.blocks = this.blocks.filter(
-        (block) => block.number <= commonAncestor.number,
-      );
-      this.logs = this.logs.filter(
-        (log) => log.blockNumber <= commonAncestor.number,
-      );
+/**
+ * Traverse the remote chain until we find a block that is
+ * compatible with out local chain.
+ *
+ * @param pendingLatestBlock Block that caused reorg to be detected.
+ * Must be at most 1 block ahead of the local chain.
+ */
+export const handleReorg = async (
+  realtimeSyncService: Service,
+  pendingLatestBlock: SyncBlock,
+) => {
+  // Prune the local chain of blocks that have been reorged out
+  realtimeSyncService.localChain = realtimeSyncService.localChain.filter(
```

Comment:
Suspicious to me that you'd mutate this at the very top of this file. What if the `deleteRealtimeData` query fails? It would make more sense to me if you use a copy, or just read the local chain at the desired height rather than mutate here.

It seems to me that the only time we mutate in the function is immediately after the `deleteRealtimeData` call.